pkg/compactor/shuffle_sharding_grouper.go

```diff
@@ -142,103 +159,132 @@ func (g *ShuffleShardingGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (re
+	for _, group := range groups {
+		var blockIds []string
+		for _, block := range group.blocks {
+			blockIds = append(blockIds, block.ULID.String())
+		}
+		blocksInfo := strings.Join(blockIds, ",")
+		level.Info(g.logger).Log("msg", "check group", "blocks", blocksInfo)
 
-// Check whether this compactor instance owns the group.
-func (g *ShuffleShardingGrouper) ownGroup(groupHash uint32) (bool, error) {
-	subRing := g.ring.ShuffleShard(g.userID, g.limits.CompactorTenantShardSize(g.userID))
+		// Nothing to do if we don't have at least 2 blocks.
+		if len(group.blocks) < 2 {
+			continue
+		}
 
-	rs, err := subRing.Get(groupHash, RingOp, nil, nil, nil)
-	if err != nil {
-		return false, err
-	}
+		groupHash := hashGroup(g.userID, group.rangeStart, group.rangeEnd)
+
+		if isLocked, err := g.isGroupLocked(group.blocks); err != nil {
+			level.Warn(g.logger).Log("msg", "unable to check if blocks in group are locked", "group hash", groupHash, "err", err, "group", group.String())
+			continue
+		} else if isLocked {
+			level.Info(g.logger).Log("msg", "skipping group because at least one block in group are locked", "group_hash", groupHash)
+			continue
+		}
+
+		remainingCompactions++
+		groupKey := createGroupKey(groupHash, group)
+
+		level.Info(g.logger).Log("msg", "found compactable group for user", "group_hash", groupHash, "group", group.String())
+		LockBlocks(g.ctx, g.bkt, g.logger, group.blocks, g.ringLifecyclerID, g.blockLockWriteFailed)
+
+		// All the blocks within the same group have the same downsample
+		// resolution and external labels.
+		resolution := group.blocks[0].Thanos.Downsample.Resolution
+		externalLabels := labels.FromMap(group.blocks[0].Thanos.Labels)
+
+		thanosGroup, err := compact.NewGroup(
+			log.With(g.logger, "groupKey", groupKey, "rangeStart", group.rangeStartTime().String(), "rangeEnd", group.rangeEndTime().String(), "externalLabels", externalLabels, "downsampleResolution", resolution),
+			g.bkt,
+			groupKey,
+			externalLabels,
+			resolution,
+			false, // No malformed index.
+			true,  // Enable vertical compaction.
+			g.compactions.WithLabelValues(groupKey),
+			g.compactionRunsStarted.WithLabelValues(groupKey),
+			g.compactionRunsCompleted.WithLabelValues(groupKey),
+			g.compactionFailures.WithLabelValues(groupKey),
+			g.verticalCompactions.WithLabelValues(groupKey),
+			g.garbageCollectedBlocks,
+			g.blocksMarkedForDeletion,
+			g.blocksMarkedForNoCompact,
+			g.hashFunc,
+			g.blockFilesConcurrency,
+			g.blocksFetchConcurrency,
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "create compaction group")
+		}
+
+		for _, m := range group.blocks {
+			if err := thanosGroup.AppendMeta(m); err != nil {
+				return nil, errors.Wrap(err, "add block to compaction group")
+			}
+		}
 
-	if len(rs.Instances) != 1 {
-		return false, fmt.Errorf("unexpected number of compactors in the shard (expected 1, got %d)", len(rs.Instances))
+		outGroups = append(outGroups, thanosGroup)
+		break mainLoop
```

Comment:
I think we should return the concurrency number of groups as the compactor can be configured to compact more than one group per time.